File: 2406-Divide Intervals Into Minimum Number of Groups.py
from heapq import heappush, heappop
class Solution:
    def minGroups(self, intervals: List[List[int]]) -> int:
        
        res = 0
        intervals.sort(key = lambda x:x[0])
        pq = []
        
        for start, end in intervals:
            if not pq or start <= pq[0]:
                res += 1
            else:
                heappop(pq)
            heappush(pq, end)
                
        return res

File: C++/2406-Divide Intervals Into Minimum Number of Groups.cpp
class Solution {
public:
    int minGroups(vector<vector<int>>& intervals) {
        sort(intervals.begin(), intervals.end());
        priority_queue<int, vector<int>, greater<>> pq;

        for (auto &interval:intervals) {
            int u = interval[0], v = interval[1];
            if (pq.empty() || u <= pq.top()) {
                pq.push(v);
            } else {
                pq.pop();
                pq.push(v);
            }
        }

        return pq.size();
    }
};